src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -310,7 +315,24 @@ private AnalyzerConfigDocumentState GetRequiredAnalyzerConfigDocumentState(Docum
         => GetRequiredProjectState(documentId.ProjectId).AnalyzerConfigDocumentStates.GetRequiredState(documentId);
 
     public ProjectState? GetProjectState(ProjectId projectId)
-        => ProjectStates.TryGetValue(projectId, out var state) ? state : null;
+        => GetProjectState(ProjectStates, projectId);
+
+    /// <summary>
+    /// Searches for the project state with the specified project id in the given project states.
+    /// </summary>
+    /// <remarks>Requires the input array to be sorted by Id</remarks>
+    private static ProjectState? GetProjectState(ImmutableArray<ProjectState> sortedPojectStates, ProjectId projectId)
+    {
+        var index = sortedPojectStates.BinarySearch(projectId, static (projectState, projectId) => CompareProjectIds(projectState.Id, projectId));
+
+        return index >= 0 ? sortedPojectStates[index] : null;
+    }
+
+    internal static int CompareProjectStates(ProjectState projectState1, ProjectState projectState2)
+        => CompareProjectIds(projectState1.Id, projectState2.Id);
+
+    private static int CompareProjectIds(ProjectId projectId1, ProjectId projectId2)
+        => projectId1.Id.CompareTo(projectId2.Id);
```

Comment:
should we make ProjectState itself IComparable?

Reply:
ProjectState: 
  yep, will go ahead and do that

ProjectId:
  this class implements IComparable, but it doesn't implement the method publicly. So, it would require casts to IComparable to be able to use it. It seemed easier to just define the method, but I can do the casts or change the interface impl to be public if you prefer.

Reply:
I'll just go ahead and cast it